src/backend/gl/src/device.rs

```diff
@@ -1744,9 +1795,21 @@ pub(crate) fn wait_fence(fence: &n::Fence, share: &Starc<Share>, timeout_ns: u64
     let gl = &share.context;
     unsafe {
         if share.private_caps.sync {
+            let sync = match fence.0.get() {
+                n::FenceState::Raw(sync) => sync,
+                n::FenceState::Unsignaled => {
+                    let sync = gl.fence_sync(glow::SYNC_GPU_COMMANDS_COMPLETE, 0).unwrap();
```

Comment:
This seems wrong. I read it as "if user wants to wait for a fence that has Unsignaled state, we just wait for GPU to complete the operation". This is wrong because what is desired here is to wait for the signal state of the fence, not for GPU to complete commands. Fences can be scheduled to signal after submission or image acquisition. So these are the places were I'd expect `gl.fence_sync()` call to be inserted, and the fence to be initialized with the `Raw` variant.

If we follow this logic, in `wait_fence` we expect the fence to be either `Raw` or `Signaled`, since otherwise with `Unsignaled` there is no ability for it to ever turn to signaled. Interestingly, this matches Metal backend logic: https://github.com/gfx-rs/gfx/blob/a0dfe9ca10c28437875081ddf45cbe6a561cc02e/src/backend/metal/src/device.rs#L2509

As a side note, you can follow the Metal backend and declare the fence as:
```rust
pub enum FenceInner {
    Idle { signaled: bool },
    Pending(<GlContext as glow::Context>::Fence),
}
```

Reply:
Aha, yeah I wasn't thinking about this correctly, and I like the way the Metal backend does this so I'll follow that pattern.

So, okay what state should the fence be after being reset?  Should it be a new unsignaled raw fence?  Why does the metal API have an `Idle { signaled: false }` state?